shap/explainers/_tree.py

```diff
@@ -1786,39 +1772,43 @@ def __init__(self, xgb_model):
-            self.node_cright.append(np.zeros(self.num_nodes[i], dtype=np.int32))
-            self.node_sindex.append(np.zeros(self.num_nodes[i], dtype=np.uint32))
-            self.node_info.append(np.zeros(self.num_nodes[i], dtype=np.float32))
-            for j in range(self.num_nodes[i]):
-                self.node_parents[-1][j] = self.read('i')
-                self.node_cleft[-1][j] = self.read('i')
-                self.node_cright[-1][j] = self.read('i')
-                self.node_sindex[-1][j] = self.read('I')
-                self.node_info[-1][j] = self.read('f')
+            self.node_parents.append(np.array(tree_json["parents"], dtype=np.int32))
+            self.node_cleft.append(np.array(tree_json["left_children"], dtype=np.int32))
+            self.node_cright.append(np.array(tree_json["right_children"], dtype=np.int32))
+            self.node_sindex.append(np.array(tree_json["split_indices"], dtype=np.uint32))
+            self.node_info.append(np.array(tree_json["split_conditions"], dtype=np.float32))
 
             # load the stat nodes
-            self.loss_chg.append(np.zeros(self.num_nodes[i], dtype=np.float32))
-            self.sum_hess.append(np.zeros(self.num_nodes[i], dtype=np.float64))
-            self.base_weight.append(np.zeros(self.num_nodes[i], dtype=np.float32))
-            self.leaf_child_cnt.append(np.zeros(self.num_nodes[i], dtype=int))
-            for j in range(self.num_nodes[i]):
-                self.loss_chg[-1][j] = self.read('f')
-                self.sum_hess[-1][j] = self.read('f')
-                self.base_weight[-1][j] = self.read('f')
-                self.leaf_child_cnt[-1][j] = self.read('i')
+            self.loss_chg.append(np.array(tree_json["loss_changes"], dtype=np.float32))
+            self.sum_hess.append(np.array(tree_json["sum_hessian"], dtype=np.float64))
+            self.base_weight.append(np.array(tree_json["base_weights"], dtype=np.float32))
+            self.leaf_child_cnt.append(np.array(tree_json["default_left"], dtype=int))
+
+    @staticmethod
+    def read_xgb_params(xgb_model) -> Dict[str, Any]:
+        import xgboost
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            if version.parse(xgboost.__version__) >= version.parse("1.6.0"):
+                tmp_file = os.path.join(tmp_dir, "model.ubj")
+                xgb_model.save_model(tmp_file)
+                xgb_params = decode_ubjson_buffer(open(tmp_file, 'rb'))
+            else:
+                warnings.warn("You are using an XGBoost version below 1.6.0 which is not fully supported by shap. "
+                              "Shap falls back to encoding the model as JSON which can lead to numerical precision issues. "
+                              "Please consider upgrading to XGBoost 1.6.0 or higher.")
+                tmp_file = os.path.join(tmp_dir, "model.json")
+                xgb_model.save_model(tmp_file)
+                xgb_params = json.load(open(tmp_file))
```

Comment:
Probably nothing, but I think it's safer to use a `with` statement here to ensure the file handle is closed properly before the temporary directory is deleted. I've seen some unexpected errors in the past on different OSs relating to file handle behaviour, so probably worth it to avoid surprises.

```suggestion
                with open(tmp_file) as fh:
                    xgb_params = json.load(fh)
```

Reply:
@CloseChoice As this is a v minor change I'll go ahead and make this tweak so we can complete the PR :)